apsuite/commisslib/measure_tbt_data.py

```diff
@@ -0,0 +1,217 @@
+            popt, *_ = _curve_fit(f=self.harmonic_tbt_model,
+                                  xdata=ilist,ydata=bpm_data,
+                                  p0=p0)
+            params[:, bpm_idx] = popt
+        return params
+
+    def calculate_betafunc_and_action(self, amplitudes, nominal_beta):
+        """Calculates beta function and betatron action as in Eq. (9)
+            of Ref. [1]"""
+        action = _np.sum(amplitudes**4)
+        action /= _np.sum(amplitudes**2 * nominal_beta)
+        beta = amplitudes**2/action
+        return beta, action
+
+class TbTDataParams(_AcqBPMsSignalsParams):
+    """."""
+
+    def __init__(self, hkick=None, vkick=None):
+        super().__init__()
+        self.nrpoints_before = 100
+        self.nrpoints_after = 2000
+        self.acq_rate = 'TbT'
+        self.signals2acq = 'XYS'
+        self._hkick = hkick
+        self._vkick = vkick
+
+
+    def __str__(self):
+        stg = super().__str__()
+        ftmp = '{0:26s} = {1:9.6f}  {2:s}\n'.format
+        stmp = '{0:26s} = {1:9}  {2:s}\n'.format
+        if self.hkick is None:
+            stg += stmp('hkick', 'same', '(current value will not be changed)')
+        else:
+            stg += ftmp('hkick', self.hkick, '[urad]')
+        if self.hkick is None:
+            stg += stmp('vkick', 'same', '(current value will not be changed)')
+        else:
+            stg += ftmp('vkick', self.vkick, '[urad]')
+        return stg
+
+    @property
+    def hkick(self):
+        """."""
+        return self._hkick
+
+    @hkick.setter
+    def hkick(self, val):
+        self._hkick = val
+
+    @property
+    def vkick(self):
+        """."""
+        return self._vkick
+
+    @vkick.setter
+    def vkick(self, val):
+        self._vkick = val
+
+class MeasureTbTData(_AcqBPMsSignals):
+    """."""
+    def __init__(self, filename='', isonline=False):
+        """."""
+        self.params = TbTDataParams()
+        self.isonline = isonline
+        self._fname = filename
+
+        if self.isonline:
+            self.create_devices()
+
+    def create_devices(self):
+        super().create_devices()
+        self.devices['pingh'] = PowerSupplyPU(
+            PowerSupplyPU.DEVICES.SI_INJ_DPKCKR)
+        self.devices['pinghv'] = PowerSupplyPU(PowerSupplyPU.DEVICES.SI_PING_V)
+
+    @property
+    def fname(self):
+        """."""
+        return self._fname
+
+    @fname.setter
+    def fname(self, val):
+        self._fname = val
+
+    def prepare_pingers(self):
+        """."""
+        pingh, pingv = self.devices['pingh'], self.devices['pingv']
+        hkick, vkick = self.params['hkick'], self.params['vkick']
+        pingh.strength = hkick / 1e3   # [urad]
+        pingv.strength = vkick / 1e3   # [urad]
+        # MISSING: set to listen to the same event as BPMs
+
+    def do_measurement(self):
+        currinfo = self.devices['currinfo']
+        self.prepare_pingers()
+        current_before = currinfo.current()
+        self.acquire_data()
```

Comment:
Maybe just leave a comment briefly describing important data that is stored here from `AcqBPMsSignals`. I would suggest to measure the tunes but then I saw that is already measured in the parent class.